File: Algorithms/p232_Implement_Queue_using_Stacks/p232_Implement_Queue_using_Stacks.py
class Queue(object):
    def __init__(self):
        """
        initialize your data structure here.
        """
        self.stack_in = []
        self.stack_out = []

    def test_and_shift(self):
        if not self.stack_out:
            while self.stack_in:
                self.stack_out.append(self.stack_in[-1])
                self.stack_in.pop()

    def push(self, x):
        """
        :type x: int
        :rtype: nothing
        """
        self.stack_in.append(x)

    def pop(self):
        """
        :rtype: nothing
        """
        self.test_and_shift()
        self.stack_out.pop()

    def peek(self):
        """
        :rtype: int
        """
        self.test_and_shift();
        return self.stack_out[-1]

    def empty(self):
        """
        :rtype: bool
        """
        return not self.stack_out and not self.stack_in

File: Algorithms/p232_Implement_Queue_using_Stacks/p232_Implement_Queue_using_Stacks.cpp
class Queue {
    stack<int,vector<int>> stack_in;
    stack<int,vector<int>> stack_out;
    
    void test_and_shift() {
        if(stack_out.empty()){
            while(!stack_in.empty()){
                stack_out.push(stack_in.top());
                stack_in.pop();
            }
        }
    }
public:
    // Push element x to the back of queue.
    void push(int x) {
        stack_in.push(x);
    }

    // Removes the element from in front of queue.
    void pop(void) {
        test_and_shift();
        stack_out.pop();
    }

    // Get the front element.
    int peek(void) {
        test_and_shift();
        return stack_out.top();
    }

    // Return whether the queue is empty.
    bool empty(void) {
        return stack_out.empty()&&stack_in.empty();
    }
};
